src/librustdoc/clean/mod.rs

```diff
@@ -1698,11 +1717,197 @@ fn normalize<'tcx>(
+            ObjectLifetimeDefault::Arg(default) => {
+                return match *default {
+                    ty::ReStatic => None,
+                    _ => Some(Lifetime::statik()),
+                };
+            }
+            // There is no meaningful bound. Further processing is needed…
+            ObjectLifetimeDefault::Empty => Lifetime::statik(),
+        },
+        ty::ReEarlyBound(ty::EarlyBoundRegion { name: region, .. })
+        | ty::ReLateBound(_, ty::BoundRegion { kind: ty::BrNamed(_, region), .. }) => {
+            match default {
+                ObjectLifetimeDefault::Static => return Some(Lifetime::statik()), // FIXME: This branch is not tested!
+                // > If there is more than one bound from the containing type then an explicit bound must be specified
+                // Due to ambiguity there is no default trait-object lifetime and thus elision is impossible.
+                // Don't elide the lifetime.
+                ObjectLifetimeDefault::Ambiguous => return Some(Lifetime(region)),
+                ObjectLifetimeDefault::Arg(default) => {
+                    return match *default {
+                        ty::ReEarlyBound(default) if region == default.name => None,
+                        ty::ReLateBound(
+                            _,
+                            ty::BoundRegion { kind: ty::BrNamed(_, default), .. },
+                        ) if region == default => None,
+                        _ => Some(Lifetime(region)),
+                    };
+                }
+                // There is no meaningful bound. Further processing is needed…
+                ObjectLifetimeDefault::Empty => Lifetime(region),
+            }
+        }
+        // We don't know how to handle them here if these cases are actually reachable. Just hide the lifetime.
+        ty::ReLateBound(..)
+        | ty::ReFree(..)
+        | ty::ReVar(..)
+        | ty::RePlaceholder(..)
+        | ty::ReErased
+        | ty::ReError(_) => return None,
+    };
+
+    // > If neither of those rules apply, then the bounds on the trait are used:
+    // FIXME(fmease):  Do we have tests for the substitution logic below?
+    let mut trait_regions: Vec<_> = tcx
+        .predicates_of(trait_)
+        .predicates
+        .iter()
+        .filter_map(|(pred, _)| {
+            // Look for bounds of the form `Self: 'a` for any region `'a`.
+            // FIXME(fmease): Don't use `skip_binder` (but use `remap_binder`?)
+            if let ty::PredicateKind::Clause(ty::Clause::TypeOutlives(ty::OutlivesPredicate(ty, region))) = pred.kind().skip_binder()
+                && let ty::Param(param) = ty.kind()
+                && param.name == kw::SelfUpper
+            {
+                Some(ty::subst::EarlyBinder(region).subst(tcx, tcx.mk_substs_trait(ty, substs)))
+            } else {
+                None
+            }
+        })
+        .collect();
```

Comment:
Do you have a way of using `rustc_trait_selection::traits::wf::object_region_bounds` like astconv does?